content/advent-2018/llvm-ir-and-go.md

```diff
@@ -0,0 +1,311 @@
+block_1:
+    %tmp = load i32, i32* @foo
+    %result = mul i32 %tmp, 2
+    ret i32 %result
+
+; Basic block with zero non-branching instructions and a return terminator.
+block_2:
+    ret i32 0
+}
+```
+
+#### Basic block
+
+A [basic block](https://en.wikipedia.org/wiki/Basic_block) is a sequence of zero or more non-branching instructions followed by a branching instruction (referred to as the terminator instruction). The key idea behind a basic block is that if a single instruction of the basic block is executed, then all instructions of the basic block are executed. This notion simplifies control flow analysis.
+
+#### Instruction
+
+An instruction is a non-branching LLVM IR instruction, usually performing a computation or accessing memory (e.g. [add](https://llvm.org/docs/LangRef.html#add-instruction), [load](https://llvm.org/docs/LangRef.html#load-instruction)), but not changing the control flow of the program.
+
+#### Terminator instruction
+
+A [terminator instruction](https://llvm.org/docs/LangRef.html#terminator-instructions) is at the end of each basic block, and determines where to transfer control flow once the basic block finishes executing. For instance [ret](https://llvm.org/docs/LangRef.html#ret-instruction) terminators returns control flow back to the caller function, and [br](https://llvm.org/docs/LangRef.html#br-instruction) terminators branches control flow either conditionally or unconditionally.
+
+### Static Single Assignment form
+
+One very important property of LLVM IR is that it is in [SSA](https://sv.wikipedia.org/wiki/SSA_(static_single_assignment_form))-form (Static Single Assignment), which essentially means that each register is assigned exactly once. This property simplifies data flow analysis.
+
+To handle variables that are assigned more than once in the original source code, a notion of [phi](https://llvm.org/docs/LangRef.html#phi-instruction) instructions are used in LLVM IR. A `phi` instruction essentially returns one value from a set of incoming values, based on the control flow path taken during execution to reach the phi instruction. Each incoming value is therefore associated with a predecessor basic block.
+
+For a concrete example, consider the following LLVM IR function.
+
+```llvm
+define i32 @f(i32 %a) {
+; <label>:0
+    switch i32 %a, label %default [
+        i32 42, label %case1
+    ]
+
+case1:
+    %x.1 = mul i32 %a, 2
+    br label %ret
+
+default:
+    %x.2 = mul i32 %a, 3
+    br label %ret
+
+ret:
+    %x.0 = phi i32 [ %x.2, %default ], [ %x.1, %case1 ]
+    ret i32 %x.0
+}
+```
+
+The `phi` instruction (sometimes referred to as `phi` nodes) in the above example essentially models the set of possible incoming values as distinct assignment statements, exactly one of which is executed based on the control flow path taken to reach the basic block of the `phi` instruction during execution. One way to illustrate the corresponding data flow is as follows:
```

Comment:
Possibly important addition: almost all local variables - even local variables that are modified - can be transformed into SSA form with phi nodes. AFAIK only variables of which the address is taken must be stored in an alloca.

Reply:
> Possibly important addition: almost all local variables - even local variables that are modified - can be transformed into SSA form with phi nodes. AFAIK only variables of which the address is taken must be stored in an alloca.

Definitely agree. I've added a paragraph to detail the recommendations on how to handle local variables in front-ends, as described in https://www.cs.cmu.edu/afs/cs/academic/class/15745-s13/public/lectures/L6-LLVM-Detail-1up.pdf (slide 16, "The mem2reg transform pass").